File: src/main/java/tw/racket/racketclient/MessageCollector.java
package tw.racket.racketclient;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.typeadapters.RuntimeTypeAdapterFactory;
import com.intellij.psi.PsiFile;
import org.jetbrains.annotations.NotNull;

import java.io.*;

public class MessageCollector {
    private MessageCollector() {
    }

    private static MessageCollector instance = null;

    public static MessageCollector get() {
        if (instance == null) {
            instance = new MessageCollector();
        }
        return instance;
    }

    public @NotNull Message[] messages(PsiFile file) throws IOException, InterruptedException {
        ProcessBuilder processBuilder = new ProcessBuilder();
        processBuilder.command("racket-service", file.getVirtualFile().getPath());
        processBuilder.directory(new File(file.getVirtualFile().getParent().getPath()));
        Process process = processBuilder.start();
        process.waitFor();
        InputStream jsonOutput = process.getInputStream();
        BufferedReader reader = new BufferedReader(new InputStreamReader(jsonOutput));
        StringBuilder builder = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) {
            builder.append(line);
        }
        return this.gson().fromJson(builder.toString(), Message[].class);
    }

    @NotNull
    private Gson gson() {
        return new GsonBuilder().registerTypeAdapterFactory(this.typeFactory()).create();
    }

    private RuntimeTypeAdapterFactory<Message> typeFactory() {
        return RuntimeTypeAdapterFactory
                .of(Message.class, "type")
                .registerSubtype(UnusedRequire.class)
                .registerSubtype(NewDefinition.class)
                .registerSubtype(Ignore.class)
                .registerSubtype(OnHoverInfo.class);
    }
}

File: src/main/kotlin/tw/racket/racketclient/MessageCollector.kt
package tw.racket.racketclient

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.typeadapters.RuntimeTypeAdapterFactory
import com.intellij.psi.PsiFile
import java.io.BufferedReader
import java.io.File
import java.io.IOException
import java.io.InputStreamReader

class MessageCollector private constructor() {
    fun messages(file: PsiFile): Array<Message> {
        val processBuilder = ProcessBuilder()
        processBuilder.command("racket", "-l", "racket-service", file.virtualFile.path)
        processBuilder.directory(File(file.virtualFile.parent.path))
        val process = processBuilder.start()
        process.waitFor()
        val jsonOutput = process.inputStream
        val reader = BufferedReader(InputStreamReader(jsonOutput))
        val builder = StringBuilder()
        var line: String?
        while (reader.readLine().also { line = it } != null) {
            builder.append(line)
        }
        return gson().fromJson(builder.toString(), Array<Message>::class.java)
    }

    private fun gson(): Gson {
        return GsonBuilder().registerTypeAdapterFactory(typeFactory()).create()
    }

    private fun typeFactory(): RuntimeTypeAdapterFactory<Message> {
        return RuntimeTypeAdapterFactory
                .of(Message::class.java, "type")
                .registerSubtype(UnusedRequire::class.java)
                .registerSubtype(NewDefinition::class.java)
                .registerSubtype(Ignore::class.java)
                .registerSubtype(OnHoverInfo::class.java)
    }

    companion object {
        private var instance: MessageCollector? = null
        fun get(): MessageCollector? {
            if (instance == null) {
                instance = MessageCollector()
            }
            return instance
        }
    }
}